### PR TITLE
Allow ingesting native histograms

### PR DESCRIPTION
**Pull request by yeya24**
**What this PR does**:

Allow ingesting native histograms. Feature can be enabled in Ingester via `--blocks-storage.tsdb.enable-native-histograms` flag.

Note: Not all ingestion and query tests are finished for Native histograms. I plan to create another PR to control the size of the PR itself

**Which issue(s) this PR fixes**:
Fixes #<issue number>

**Checklist**
- [x] Tests updated
- [x] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: docs/blocks-storage/querier.md [hunk 1]
```diff
@@ -1441,4 +1441,8 @@ blocks_storage:
     # be out-of-order.
     # CLI flag: -blocks-storage.tsdb.out-of-order-cap-max
     [out_of_order_cap_max: <int> | default = 32]
+
+    # True to enable native histogram.
+    # CLI flag: -blocks-storage.tsdb.enable-native-histograms
+    [enable_native_histograms: <boolean> | default = false]
```
Comment (alanprot) [inline docs/blocks-storage/querier.md:1447]: Should we mark as experimental for the time being?

Comment (yeya24) [reply docs/blocks-storage/querier.md]: Updated the flag description.

File: docs/blocks-storage/querier.md [hunk 2]
```diff
@@ -1441,4 +1441,8 @@ blocks_storage:
     # be out-of-order.
     # CLI flag: -blocks-storage.tsdb.out-of-order-cap-max
     [out_of_order_cap_max: <int> | default = 32]
+
+    # [EXPERIMENTAL] True to enable native histogram.
```
Comment (friedrichg) [inline docs/blocks-storage/querier.md:1445]: we should update https://github.com/cortexproject/cortex/blob/master/docs/configuration/v1-guarantees.md too

**Review by friedrichg — APPROVED** [review]
Amazing!. I can't wait to try this out

**Comment by yeya24** [comment]
> Amazing!. I can't wait to try this out

Note that this part is just ingestion and histograms are not queryable from Query Frontend at this point. It will be added in future PRs.

File: pkg/ingester/ingester.go [hunk 3]
```diff
@@ -1174,6 +1173,89 @@ func (i *Ingester) Push(ctx context.Context, req *cortexpb.WriteRequest) (*corte
...
+
+				// The error looks an issue on our side, so we should rollback
+				if rollbackErr := app.Rollback(); rollbackErr != nil {
+					level.Warn(logutil.WithContext(ctx, i.logger)).Log("msg", "failed to rollback on error", "user", userID, "err", rollbackErr)
+				}
+				return nil, wrapWithUser(err, userID)
+			}
+		} else {
+			nativeHistogramCount += len(ts.Histograms)
+		}
```
Comment (alanprot) [inline pkg/ingester/ingester.go:1257]: This seems duplicated? Can we extract to a new function?

Comment (yeya24) [reply pkg/ingester/ingester.go]: Done

**Comment by yeya24** [comment]
@alanprot Can you please take another look?

**Review by alanprot — APPROVED** [review]

**Pull request merged**